packages/electric-db-collection/src/electric.ts

```diff
@@ -994,14 +1241,41 @@ function createElectricSync<T extends Row<unknown>>(
                 transactionStarted = true
               }
 
-              write({
-                type: message.headers.operation,
-                value: message.value,
-                // Include the primary key and relation info in the metadata
-                metadata: {
-                  ...message.headers,
-                },
-              })
+              // Process tags if present
+              const tags = message.headers.tags
+              const removedTags = message.headers.removed_tags
+              const hasTags = tags || removedTags
+
+              const rowId = collection.getKeyFromItem(message.value)
+              const rowTagSet = () =>
+                processTagsForChangeMessage(tags, removedTags, rowId)
+
+              // Check if row should be deleted (empty tag set)
+              // but only if the message includes tags
+              // because shapes without subqueries don't contain tags
+              // so we should keep those around
+              if (hasTags && rowTagSet().size === 0) {
```

Comment:
I'm not sure this branch is possible on a change message, because an update that would have removed all tags from a row means it no longer matches the where clause and thus would be converted to a delete on the server